Modbus TCP: make port configurable

anbie opened the pull request:
This pull request makes the Modbus TCP port configurable in the lambda-zewotherm charger template.

Motivation:
- Until now, the lambda-zewotherm template hardcoded port 502 for Modbus TCP.
- Many Modbus devices accept only a single client connection. To enable multiple clients, modbus-proxies can be used (e.g., In my environment I'm using [tiagocoutinho/modbus-proxy](https://github.com/tiagocoutinho/modbus-proxy)). This proxy serializes requests by acting as a bridge. For targeting this proxy is listening on a non-standard, configurable port.

Change:
- Replaced the hardcoded 502 with a configurable port option in: `templates/definition/charger/lambda-zewotherm.yaml`
- Default port remains 502, ensuring full backward compatibility.

The change has been unit-tested in my home environment, including with a Lambda heat pump.

anbie committed:
feat: Add support for negative excess energy. Also enable port at URI to be specified

anbie committed:
Address review comments by @andig. Revert port config elements

anbie committed:
fix EncodeFunc() to enable negative numbers in writemultiple. Fix proposed by @andig

anbie committed:
Merge branch 'master' into feat/lambda_neg_excess_energy

anbie committed:
Merge branch 'master' into feat/lambda_neg_excess_energy

anbie committed:
Merge branch 'master' into feat/lambda_neg_excess_energy

anbie committed:
Update lambda-zewotherm.yaml

Removed 2 trailing spaces in line 36. I missed them before 🙄

anbie committed:
support configurable Modbus TCP port (default 502) for lambda heatpump in template

anbie committed:
Merge branch 'master' into feature/config_ports4modbustcp

anbie commented:
@andig, ich hatte diese PR auf Basis von https://github.com/evcc-io/evcc/pull/21972 erstellt.
Da in beiden PRs zwar dieselbe Datei geändert wird, aber in unterschiedlichen Zeilen, war das eher unnötig und sorgt sicher für das Durcheinander. 
Was schlägst du vor? PR schließen und Neuen öffnen?

anbie commented:
2 branches: 

#21972: feat/lambda_neg_excess_energy (https://github.com/anbie/evcc/tree/feat/lambda_neg_excess_energy)
#22196: feature/config_ports4modbustcp (https://github.com/anbie/evcc/tree/feature/config_ports4modbustcp)

andig commented:
…dennoch sind hier dinge drin die nicht rein gehören. Git revert der falschen commits?

anbie committed:
Merge branch 'master' into feature/config_ports4modbustcp

anbie committed:
Allow configuring Modbus TCP port (default 502) for lambda-zewotherm.
Sorted out branch problems ("durcheinander"). Rebased this branch onto master.
modified:   templates/definition/charger/lambda-zewotherm.yaml

anbie committed:
revert register.go to match master

anbie committed:
Merge branch 'master' into feature/config_ports4modbustcp

premultiply committed:
use modbus

anbie commented:
Ist der PR soweit wieder auf "ready for review" zu gehen ?

anbie commented:
Ich selbst habe keine Möglichkeit, Modbus-Geräte in größerem Umfang zu testen. Wenn ich es richtig gesehen habe, lässt sich der Port aber bei den meisten Charger ohnehin schon konfigurieren – nur bei wenigen ist er noch hart auf 502 gesetzt. Aber ich lass hier mal gut sein und konfiguriere meinen modbus proxy anders.

anbie committed:
Merge branch 'master' into feature/config_ports4modbustcp

anbie commented:
Ich verstehe deine Bedenken – bestehende Konfigurationen sollten auf keinen Fall gebrochen werden. Deshalb war mein initialer Vorschlag vor 3 Wochen der hier:

```
params:
  - name: host
  - name: port
    default: 502
[...]
render: |
  type: heatpump
  setmaxpower:
    source: watchdog
    timeout: {{ .watchdog }} # re-write at timeout/2
    initial: 0
    set:
      source: modbus
      uri: {{ .host }}:{{ .port }}
      id: 1
[...]
```
Das ist nicht so vielseitig und elegant wie der Ansatz von @premultiply, sollte aber nach meinem Verständniss die "backward compabilbility" sicherstellen. Oder übersehe ich hier etwas? @andig

anbie committed:
Merge branch 'master' into feature/config_ports4modbustcp

andig commented:
> Deshalb war mein initialer Vorschlag vor 3 Wochen der hier:

Es tut mir leid- dann lass uns dahin zurück gehen. Sorry für die Extrarunde.

anbie committed:
Revert to optional port parameter in lambda‑zewotherm template

anbie committed:
add in id: 1

anbie committed:
add sponsortoken line

anbie commented:
Danke @andig!
